server/comms/server.go

```diff
@@ -452,3 +553,39 @@ func parseListeners(addrs []string) ([]string, []string, bool, error) {
 	}
 	return ipv4ListenAddrs, ipv6ListenAddrs, haveWildcard, nil
 }
+
+// handleHTTP handles an HTTP request by finding and calling the appropriate
+// handler. The caller of handleHTTP should have already parsed and request
+// data into the right type for the handler.
+func handleHTTP(w http.ResponseWriter, r *http.Request, route string, thing interface{}) {
+	handler := httpRoutes[route]
+	if handler == nil {
+		http.Error(w, http.StatusText(http.StatusNotFound), http.StatusNotFound)
+		return
+	}
+	resp, err := handler(thing)
+	if err != nil {
+		writeJSONWithStatus(w, map[string]string{"error": err.Error()}, http.StatusBadRequest)
+		return
+	}
+	writeJSONWithStatus(w, resp, http.StatusOK)
+}
+
+// routeHandler creates a HandlerFunc for a route. Middleware should have
+// already processed the request and added the request struct to the Context.
+func routeHandler(route string) func(w http.ResponseWriter, r *http.Request) {
+	return func(w http.ResponseWriter, r *http.Request) {
+		handleHTTP(w, r, route, r.Context().Value(ctxThing))
+	}
+}
+
+// writeJSONWithStatus writes the JSON response with the specified HTTP response
+// code.
+func writeJSONWithStatus(w http.ResponseWriter, thing interface{}, code int) {
+	w.Header().Set("Content-Type", "application/json; charset=utf-8")
+	w.WriteHeader(code)
+	encoder := json.NewEncoder(w)
+	if err := encoder.Encode(thing); err != nil {
```

Comment:
An issue I ran into with this that if their is a json marshalling error, it still sends junk to the client.  https://github.com/decred/dcrdex/commit/ae3b4ff45adcd9614aa175caf0bb1acc825af672
If we are positive `thing` will always marshal without error, it's OK though.  Plus we're following the NewEncoder pattern in all the other http handlers and packages, so maybe we stay consistent with them for now.